src/fairseq2/datasets/preference.py

```diff
@@ -201,66 +204,142 @@ def create_reader(
-
-        builder.map(source_encoder, selector="src")
-        builder.map(target_encoder, selector="tgt_chosen")
-        builder.map(target_encoder, selector="tgt_rejected")
+        if options.chat_mode is True:
+            if not isinstance(
+                getattr(tokenizer, "_model", None), HuggingFaceTokenModel
+            ):
+                raise RuntimeError(
+                    "Huggingface tokenizer must be used when chat_mode is True"
+                )
 
-        def cat_source_and_target(example: dict[str, Any]) -> dict[str, Any]:
-            id_ = example.get("id", None)
+            def encoding_chat(example: dict[str, Any]) -> dict[str, Any]:
+                id_ = example.get("id")
+                chat_chosen = example.get("chat_chosen")
+                chat_rejected = example.get("chat_rejected")
 
-            source_indices = example["src"]
-            target_indices_chosen = example["tgt_chosen"]
-            target_indices_rejected = example["tgt_rejected"]
+                encoded_output_chosen = tokenizer._model._tok.apply_chat_template(
+                    chat_chosen,
+                    return_dict=True,
+                    return_assistant_tokens_mask=True,
+                    return_tensors="pt",
+                )
 
-            indices_chosen = torch.cat([source_indices, target_indices_chosen])
-            indices_rejected = torch.cat([source_indices, target_indices_rejected])
+                indices_chosen = encoded_output_chosen["input_ids"][0]
+                target_mask_chosen = encoded_output_chosen["assistant_masks"][0].bool()
 
-            if options.mask_source_tokens:
-                source_len = len(source_indices)
-                target_mask_chosen = torch.arange(len(indices_chosen)) >= source_len
-                target_mask_rejected = torch.arange(len(indices_rejected)) >= source_len
-            else:
-                target_mask_chosen = torch.full([len(indices_chosen)], True)
-                target_mask_rejected = torch.full([len(indices_rejected)], True)
+                encoded_output_rejected = tokenizer._model._tok.apply_chat_template(
+                    chat_rejected,
+                    return_dict=True,
+                    return_assistant_tokens_mask=True,
+                    return_tensors="pt",
+                )
 
-            total_tokens = (
-                2 * len(source_indices)
-                + len(target_indices_chosen)
-                + len(target_indices_rejected)
-            )
+                indices_rejected = encoded_output_rejected["input_ids"][0]
+                target_mask_rejected = encoded_output_rejected["assistant_masks"][
+                    0
+                ].bool()
+
+                if not options.mask_source_tokens:
+                    # no source masking i.e. mask has all 1s
+                    target_mask_chosen._fill(True)
+                    target_mask_rejected._fill(True)
```

Comment:
this isn't in instruction.py encoding_chat() function. is that correct?

Reply:
yes, SFT doesnt have src masking support, it only exists in DPO since one team figured it helped them 😅